src/wallet/wallet_tx_builder.cpp

```diff
@@ -0,0 +1,751 @@
+                        }
+                        break;
+                    case ReceiverType::Sapling:
+                        if (!spendable.saplingNoteEntries.empty() || strategy.AllowRevealedAmounts()) {
+                            result.insert(OutputPool::Sapling);
+                        }
+                        break;
+                    case ReceiverType::Orchard:
+                        if (params.GetConsensus().NetworkUpgradeActive(anchorHeight, Consensus::UPGRADE_NU5)
+                            && (!spendable.orchardNoteMetadata.empty() || strategy.AllowRevealedAmounts())) {
+                            result.insert(OutputPool::Orchard);
+                        }
+                        break;
+                }
+            }
+            return result;
+        };
+
+        std::visit(match {
+            [&](const CKeyID& keyId) {
+                auto sendTo = pwalletMain->GenerateChangeAddressForAccount(
+                        sendFromAccount,
+                        allowedChangeTypes({ReceiverType::P2PKH}));
+                if (sendTo.has_value()) {
+                    resolvedPayments.AddPayment(
+                            ResolvedPayment(std::nullopt, sendTo.value(), changeAmount, std::nullopt, true));
+                    changeAddr = sendTo.value();
+                }
+            },
+            [&](const CScriptID& scriptId) {
+                auto sendTo = pwalletMain->GenerateChangeAddressForAccount(
+                        sendFromAccount,
+                        allowedChangeTypes({ReceiverType::P2SH}));
+                assert(sendTo.has_value());
+                resolvedPayments.AddPayment(
+                        ResolvedPayment(std::nullopt, sendTo.value(), changeAmount, std::nullopt, true));
+                changeAddr = sendTo.value();
+            },
+            [&](const libzcash::SproutPaymentAddress& addr) {
+                // for Sprout, we return change to the originating address using the tx builder.
+                changeAddr = addr;
+            },
+            [&](const libzcash::SproutViewingKey& vk) {
+                // for Sprout, we return change to the originating address using the tx builder.
+                changeAddr = vk.address();
+            },
+            [&](const libzcash::SaplingPaymentAddress& addr) {
+                // for Sapling, if using a legacy address, return change to the
+                // originating address; otherwise return it to the Sapling internal
+                // address corresponding to the UFVK.
+                std::optional<RecipientAddress> sendTo;
+                if (sendFromAccount == ZCASH_LEGACY_ACCOUNT) {
+                    sendTo = addr;
+                } else {
+                    sendTo = pwalletMain->GenerateChangeAddressForAccount(
+                            sendFromAccount,
+                            allowedChangeTypes({ReceiverType::Sapling}));
```

Comment:
`CWallet::ZTXOSelectorForAddress` only returns `AccountZTXOPattern` when `AllowLinkingAccountAddresses` is permitted (which is overly strict, but that’s not this PR).

I assume it would be possible to have it return a `UnifiedAddress` pattern in the case case when the user provides a non-legacy `SaplingPaymentAddress`, but currently when given an explicit Sapling address, it returns that address as the selector. Then `CWallet::FindAccountForSelector` does a bunch of hard-to-follow stuff … and maybe returns an `AccountId` at the end. But, considering that when given a `SaplingPaymentAddress` (or anything other than Sprout), it doesn’t immediately return `std::nullopt` leads me to believe that the `AccountId` in these cases can be non-legacy.